arm_gdb/mpu.py

```diff
@@ -0,0 +1,248 @@
+        ]),
+        RegisterDef(f"MPU_RLAR", f"MPU Region Limit Address Register", 0xE000EDA0, 4, [
+            FieldBitfield("LIMIT", 5, 27,
+                        "Limit address. Contains bits [31:5] of the upper inclusive limit of the selected MPU memory region"),
+            FieldBitfieldEnum("PXN", 4, 1, [
+                            (0b0, False, "Execution only permitted if read permitted.", None),
+                            (0b1, False, "Execution from a privileged mode is not permitted.", None),
+                        ],
+                        "Privileged execute-never. Defines whether code can be executed from this privileged region."),
+            FieldBitfield("AttrIndx", 1, 3,
+                        "Attribute index. Associates a set of attributes in the MPU_MAIR0 and MPU_MAIR1 fields."),
+            FieldBitfieldEnum("EN", 0, 1, [
+                            (0b0, False, "Region disabled.", None),
+                            (0b1, False, "Region enabled.", None),
+                        ],
+                        "Enable. Region enable."),
+        ])
+    ]
+
+def get_mpu_dregions(inf):
+    MPU_TYPE = read_reg(inf, 0xE000ED90, 4)
+    return (MPU_TYPE >> 8) & 0xff
+
+def get_mpu_region(inf):
+    MPU_RNR = read_reg(inf, 0xE000ED98, 4)
+    return (MPU_RNR >> 0) & 0xff
+
+def set_mpu_region(inf, region):
+    write_reg(inf, 0xE000ED98, region & 0xff, 4)
+
+def is_mpu_region_enabled(inf):
+    MPU_RLAR = read_reg(inf, 0xE000EDA0, 4)
+    return ((MPU_RLAR >> 0) & 1) == 1
+
+
+class ArmToolsMPU (ArgCommand):
+    """Dump of ARM Cortex-M MPU - Memory Protection Unit registers
+
+Usage: arm mpu [/habf]
+
+Modifier /h provides descriptions of names where available
+Modifier /a Print all fields, including default values and disabled regions
+Modifier /b prints bitmasks in binary instead of hex
+Modifier /f force printing fields from all Cortex-M models
+"""
+
+    def __init__(self):
+        super().__init__('arm mpu', gdb.COMMAND_DATA)
+        self.add_mod('h', 'descr')
+        self.add_mod('a', 'all')
+        self.add_mod('b', 'binary')
+        self.add_mod('f', 'force')
+
+    def invoke(self, argument, from_tty):
+        args = self.process_args(argument)
+        if args is None:
+            self.print_help()
+            return
+
+        try:
+            base = 1 if args['binary'] else 4
+
+            inf = gdb.selected_inferior()
+
+            # Detect CPU type, convert to a useful key for dicts
+            CPUID = read_reg(inf, 0xE000ED00, 4)
+            model = {
+                "4100c200": ["M0", "v6"],
+                "4100c600": ["M0+", "v6"],
+                "4100c210": ["M1", "v6"],
+                "4100c230": ["M3", "v7"],
+                "4100c240": ["M4", "v7"],
+                "4100c270": ["M7", "v7"],
+                # TODO: support ARMv8-M, for now pretend it's v7, since it's similar
+                "4100d200": ["M23", "v8"],
+                "4100d210": ["M33", "v8"],
```

Comment:
My cpuid is 
CPUID                            = 630f1321 
    Implementer                    63...... - 63             
    Variant                        ..0..... - Revision: r0pX 
    Architecture                   ...f.... - ARMv7-M        
    PartNo                         ....132. - 132            
    Revision                       .......1 - Patch: rXp1
but it is cortex-m33, and armv8m. please and this to dict?
                "63001320": ["M33", "v8"]

Reply:
Gosh... CPUID format is a mess. Apparently Architecture field of 1111 is the architecture in ARMv7-M reference manual, but in ARMv8-M it's also a valid value... Need to investigate a bit.

But what CPU are you using? Since "Implementer" 63 is not ARM, so I guess it's a clone? ARM should be 41

Reply:
I've added it for now at least

Reply:
Thank you, it is China star-mc1, a clone of cortex-m33.